File: Source/Aladdin.CAPI.Java/src/aladdin/capi/pbe/PBES2.java
package aladdin.capi.pbe;
import aladdin.*; 
import aladdin.capi.*; 
import java.security.*; 
import java.io.*; 

///////////////////////////////////////////////////////////////////////////
// Алгоритм шифрования по паролю PBES2
///////////////////////////////////////////////////////////////////////////
public class PBES2 extends Cipher
{
	private final KeyDerive derivationAlgorithm;    // алгоритм наследования
	private final Cipher    cipherAlgorithm;        // алгоритм шифрования
    
	// конструктор 
	public PBES2(KeyDerive derivationAlgorithm, Cipher cipherAlgorithm) 
	{
        // сохранить переданные параметры
		this.derivationAlgorithm = RefObject.addRef(derivationAlgorithm);
		this.cipherAlgorithm	 = RefObject.addRef(cipherAlgorithm);	
	}
    // освободить выделенные ресурсы
    @Override protected void onClose() throws IOException 
    { 
        // освободить выделенные ресурсы
        RefObject.release(derivationAlgorithm); 
        
        // освободить выделенные ресурсы
        RefObject.release(cipherAlgorithm); super.onClose();        
    } 
    // размер блока алгоритма
    @Override public int blockSize() { return cipherAlgorithm.blockSize(); }
    
	// алгоритм зашифрования данных
	@Override protected Transform createEncryption(ISecretKey password) 
        throws IOException, InvalidKeyException
	{
        // определить тип ключа
        SecretKeyFactory keyFactory = cipherAlgorithm.keyFactory(); int keySize = -1; 
        
        // определить допустимые размеры ключей
        int[] keySizes = cipherAlgorithm.keySizes(); 
        
        // указать рекомендуемый размер ключа
        if (keySizes != null && keySizes.length == 1) keySize = keySizes[0]; 
        
		// наследовать ключ по паролю
		try (ISecretKey key = derivationAlgorithm.deriveKey(password, null, keyFactory, keySize)) 
        {
            // проверить допустимость размера ключа
            if (!KeySizes.contains(keySizes, key.length())) 
            {
                // выбросить исключение
                throw new IllegalStateException();
            }
            // вернуть преобразование зашифрования
            return cipherAlgorithm.createEncryption(key, PaddingMode.PKCS5); 
        }
	}
	// алгоритм расшифрования данных
	@Override protected Transform createDecryption(ISecretKey password) 
        throws IOException, InvalidKeyException
	{
        // определить тип ключа
        SecretKeyFactory keyFactory = cipherAlgorithm.keyFactory(); int keySize = -1; 
        
        // определить допустимые размеры ключей
        int[] keySizes = cipherAlgorithm.keySizes(); 
        
        // указать рекомендуемый размер ключа
        if (keySizes != null && keySizes.length == 1) keySize = keySizes[0]; 
        
        // наследовать ключ по паролю
		try (ISecretKey key = derivationAlgorithm.deriveKey(password, null, keyFactory, keySize))
        {
            // проверить допустимость размера ключа
            if (!KeySizes.contains(keySizes, key.length())) 
            {
                // выбросить исключение
                throw new IllegalStateException();
            }
            // вернуть преобразование расшифрования
            return cipherAlgorithm.createDecryption(key, PaddingMode.PKCS5); 
        }
	}
}

File: Source/Aladdin.CAPI/CAPI/PBE/PBES2.cs
using System;

namespace Aladdin.CAPI.PBE
{
	///////////////////////////////////////////////////////////////////////////
	// Алгоритм шифрования по паролю PBES2
	///////////////////////////////////////////////////////////////////////////
    public class PBES2 : Cipher
	{
		private KeyDerive derivationAlgorithm;  // алгоритм наследования
		private Cipher    cipherAlgorithm;	    // алгоритм шифрования

        // создать алгоритм
        public static PBES2 Сreate(Factory factory, 
            SecurityStore scope, ASN1.ISO.PKCS.PKCS5.PBES2Parameter pbeParameters)
        {
            // создать алгоритм шифрования
            using (Cipher cipher = factory.CreateAlgorithm<Cipher>(
                scope, pbeParameters.EncryptionScheme))
            {
                // проверить наличие алгоритма
                if (cipher == null) return null; 
                    
                // создать алгоритм наследования ключа по паролю
                using (KeyDerive derivationAlgorithm = factory.CreateAlgorithm<KeyDerive>(
                    scope, pbeParameters.KeyDerivationFunc))
                {
                    // проверить наличие алгоритма
                    if (derivationAlgorithm == null) return null; 
                 
                    // создать алгоритм шифрования по паролю
                    return new PBES2(derivationAlgorithm, cipher);  
                }
            }
        }
		// конструктор 
		public PBES2(KeyDerive derivationAlgorithm, Cipher cipherAlgorithm)
		{
            // сохранить переданные параметры
			this.derivationAlgorithm = RefObject.AddRef(derivationAlgorithm);
			this.cipherAlgorithm	 = RefObject.AddRef(cipherAlgorithm);	
		}
        // освободить выделенные ресурсы
        protected override void OnDispose()
        {
            // освободить выделенные ресурсы
            RefObject.Release(derivationAlgorithm); 
                
            // освободить выделенные ресурсы
            RefObject.Release(cipherAlgorithm); base.OnDispose();
        }
        // размер блока алгоритма
        public override int BlockSize { get { return cipherAlgorithm.BlockSize; }}

		// алгоритм зашифрования данных
		protected override Transform CreateEncryption(ISecretKey password)
		{
            // определить тип ключа
            SecretKeyFactory keyFactory = cipherAlgorithm.KeyFactory; int keySize = -1; 

            // определить допустимые размеры ключей
            int[] keySizes = cipherAlgorithm.KeyFactory.KeySizes; 
        
            // указать рекомендуемый размер ключа
            if (keySizes != null && keySizes.Length == 1) keySize = keySizes[0]; 
        
			// наследовать ключ по паролю
			using (ISecretKey key = derivationAlgorithm.DeriveKey(password, null, keyFactory, keySize))
            { 
                // проверить допустимость размера ключа
                if (!CAPI.KeySizes.Contains(keySizes, key.Length)) 
                {
                    // выбросить исключение
                    throw new InvalidOperationException();
                }
 			    // вернуть преобразование зашифрования
			    return cipherAlgorithm.CreateEncryption(key, PaddingMode.PKCS5); 
            }
		}
		// алгоритм расшифрования данных
		protected override Transform CreateDecryption(ISecretKey password)
		{
            // определить тип ключа
            SecretKeyFactory keyFactory = cipherAlgorithm.KeyFactory; int keySize = -1; 

            // определить допустимые размеры ключей
            int[] keySizes = cipherAlgorithm.KeyFactory.KeySizes; 
        
            // указать рекомендуемый размер ключа
            if (keySizes != null && keySizes.Length == 1) keySize = keySizes[0]; 

			// наследовать ключ по паролю
			using (ISecretKey key = derivationAlgorithm.DeriveKey(password, null, keyFactory, keySize))
            { 
                // проверить допустимость размера ключа
                if (!CAPI.KeySizes.Contains(keySizes, key.Length)) 
                {
                    // выбросить исключение
                    throw new InvalidOperationException();
                }
 			    // вернуть преобразование расшифрования
			    return cipherAlgorithm.CreateDecryption(key, PaddingMode.PKCS5);
            } 
		}
	}
}
